test/sslapitest.c

```diff
@@ -656,6 +657,178 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
+
+    memcpy(crec_wseq_before, &clientssl->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    memcpy(srec_wseq_before, &serverssl->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    memcpy(srec_rseq_before, &serverssl->rlayer.read_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    if (!TEST_true(SSL_write(clientssl, cbuf, sizeof(cbuf)) == sizeof(cbuf)))
+        goto end;
+
+    while ((err = SSL_read(serverssl, &sbuf, sizeof(sbuf))) != sizeof(sbuf)) {
+        if (SSL_get_error(serverssl, err) != SSL_ERROR_WANT_READ) {
+            goto end;
+        }
+    }
+
+    if (!TEST_true(SSL_write(serverssl, sbuf, sizeof(sbuf)) == sizeof(sbuf)))
+        goto end;
+
+    while ((err = SSL_read(clientssl, &cbuf, sizeof(cbuf))) != sizeof(cbuf)) {
+        if (SSL_get_error(clientssl, err) != SSL_ERROR_WANT_READ) {
+            goto end;
+        }
+    }
+
+    memcpy(crec_wseq_after, &clientssl->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    memcpy(srec_wseq_after, &serverssl->rlayer.write_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    memcpy(srec_rseq_after, &serverssl->rlayer.read_sequence,
+            TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    /* verify the payload */
+    if (!TEST_true(memcmp(cbuf, sbuf, sizeof(cbuf)) == 0))
+        goto end;
+
+    /* ktls is used then kernel sequences are used instead of OpenSSL sequences */
+    if (clientssl->mode & SSL_MODE_NO_KTLS_TX) {
+        if (!TEST_true(memcmp(crec_wseq_before, crec_wseq_after,
+                        TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE) != 0))
+            goto end;
+    } else {
+        if (!TEST_true(memcmp(crec_wseq_before, crec_wseq_after,
+                        TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE) == 0))
+            goto end;
+    }
+
+    if (serverssl->mode & SSL_MODE_NO_KTLS_TX) {
+        if (!TEST_true(memcmp(srec_wseq_before, srec_wseq_after,
+                        TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE) != 0))
+            goto end;
+    } else {
+        if (!TEST_true(memcmp(srec_wseq_before, srec_wseq_after,
+                        TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE) == 0))
+            goto end;
+    }
+
+    if (!TEST_true(memcmp(srec_rseq_before, srec_rseq_after,
+                          TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE) != 0))
+        goto end;
+
+    return 1;
+end: 
+    return 0;
+}
+
+static int execute_test_ktls(int cis_ktls_tx, int sis_ktls_tx)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    int cfd, sfd;
+
+    if (!TEST_true(create_test_sockets(&cfd, &sfd)))
+        goto end;
+
+    /* Create a session based on SHA-256 */
+    if (!TEST_true(create_ssl_ctx_pair(TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_VERSION, TLS1_2_VERSION,
```

Comment:
Does KTLS support TLS1_VERSION? I thought it was TLSv1.2 only? In any case I think we're going to need to skip these tests if only TLSv1.3 is enabled.

Reply:
Only TLS_1_2.
Will fix.